text/0000-async-stream.md

```diff
@@ -0,0 +1,903 @@
+
+As detailed in previous sections, the migrations to add these traits are out of scope for this RFC.
+
+## Async iteration syntax
+
+Currently, if someone wishes to iterate over a `Stream` as defined in the `futures` crate,
+they are not able to use  `for` loops, they must use `while let` and `next/try_next` instead.
+
+We may wish to extend the `for` loop so that it works over streams as well. 
+
+```rust
+#[async]
+for elem in stream { ... }
+```
+
+One of the complications of using `while let` syntax is the need to pin.
+A `for` loop syntax that takes ownership of the stream would be able to
+do the pinning for you. 
+
+We may not want to make sequential processing "too easy" without also enabling
+parallel/concurrent processing, which people frequently want. One challenge is
+that parallel processing wouldn't naively permit early returns and other complex
+control flow. We could add a `par_stream()` method, similar to 
+[Rayon's](https://github.com/rayon-rs/rayon) `par_iter()`.
+
+Designing this extension is out of scope for this RFC. However, it could be prototyped using procedural macros today.
+
+## "Lending" streams
+
+There has been much discussion around lending streams (also referred to as attached streams).
+
+### Definitions
+
+[Source](https://smallcultfollowing.com/babysteps/blog/2019/12/10/async-interview-2-cramertj-part-2/#the-need-for-streaming-streams-and-iterators)
+
+
+In a **lending** stream (also known as an "attached" stream), the `Item` that gets 
+returned by `Stream` may be borrowed from `self`. It can only be used as long as 
+the `self` reference remains live.
+
+In a **non-lending** stream (also known as a "detached" stream), the `Item` that 
+gets returned by `Stream` is "detached" from self. This means it can be stored 
+and moved about independently from `self`.
+
+This RFC does not cover the addition of lending streams (streams as implemented through 
+this RFC are all non-lending streams).
+
+We can add the `Stream` trait to the standard library now and delay
+adding in this distinction between the two types of streams - lending and
+non-lending. The advantage of this is it would allow us to copy the `Stream`
+trait from `futures` largely 'as is'. 
+
+The disadvantage of this is functions that consume streams would 
+first be written to work with `Stream`, and then potentially have 
+to be rewritten later to work with `LendingStream`s.
+
+### Current Stream Trait
+
+```rust
+pub trait Stream {
+    type Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+}
+```
+
+This trait, like `Iterator`, always gives ownership of each item back to its caller. This offers flexibility - 
+such as the ability to spawn off futures processing each item in parallel.
+
+### Potential Lending Stream Trait
+
+```rust
```

Comment:
This will only allow `Item: 'static`. It should actually be 

```rust
trait LendingStream<'s> {
    type Item<'a> where 's: 'a;

    fn poll_next<'a>(
        self: Pin<&'b mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item<'a>>>;
}
```
See `Iterator` examples:
- [Not working](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4e5913d006b685f4c44f9279ee5fc0d8)
- [Working](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0801b95a13c4604dc79a80f84bfbe1e6)

Reply:
Good points! Adding.